### DISCUSSION TITLE
Hybrid Multitenancy in Reactive using Flyway

### DISCUSSION THREAD

**Discussion by humcqc** [quarkusio/quarkus#33342] [Community]
Hello,

I have a requirement to be able to store:
- small tenants with Discriminator strategy, 
- some medium tenants into different schemas, 
- and the biggest in different databases. 
And this dynamically if possible.

On the same time I would like to:
- implement my services in reactive
- use flyway for the db migration

After some discussion with @yrodiere on https://github.com/quarkusio/quarkus/discussions/32686, I would like to use this place to discuss if this is something doable , the possible design,  the current issues, and what should be done.

List of issues found:
- https://github.com/quarkusio/quarkus/issues/28644
- https://github.com/quarkusio/quarkus/issues/34443
- https://github.com/quarkusio/quarkus/issues/34071
- https://github.com/quarkusio/quarkus/issues/10716
- https://github.com/quarkusio/quarkus/issues/36581
- https://github.com/quarkusio/quarkus/issues/15959
- https://github.com/quarkusio/quarkus/issues/13425
- https://github.com/flyway/flyway/issues/1308
- https://github.com/quarkusio/quarkus/issues/21110
- https://github.com/quarkusio/quarkus/issues/28629

Following some answers in those issues, it seems there is a discussion on Reactive design , if there is some others known issues  I can subscribe to see the direction of this I will be interested. 

@yrodiere do you have some idea ?

**Comment by yrodiere** [comment]
Hi,

I don't really deal with Hibernate Reactive on a daily basis, so I'm not fully aware of what it can and cannot do.

However, I would imagine the problem is pretty close to what it would be in Hibernate ORM. As I mentioned in #32686:

> The first step would be to check whether it's possible in Hibernate ORM itself, without Quarkus. If it is, then I think we should be able to make it work simply by making the "multitenancy" configuration property multi-valued.

And the best person to ask whether such hybrid multitenancy can work (especially in Hibernate Reactive) would probably be @gavinking.

**Reply by gavinking** [reply]
I mean the latest release of HR is based on H6, which has discriminator-based multitenancy via the `@TenantId` annotation.

And for schema/catalog-based multitenancy, there's the [documented solution](https://hibernate.org/reactive/documentation/2.0/reference/html_single/#_custom_connection_management_and_multitenancy) of extending `DefaultSqlClientPool`.

[P.S. `hibernate.multiTenancy` no longer exists in H6, the HR docs still need to be updated.]

So anyway, I'm not sure what's missing.

Have you tried it?

**Reply by humcqc** [reply]
Seems blocked by https://github.com/quarkusio/quarkus/issues/15959

What do you recommend for the optimal way for multitenancy in quarkus ?
- waiting for reactive fix
- using DefaultSqlClientPool
- using virtual thread in orm way
